### PR TITLE
Basic plumbing to inspect llvm/asm from compiled functions.

### PR DESCRIPTION
**Pull request by meawoppl**
The basic code needed to implement inspect_asm() and inspect_llvm() as discussed in issue #837 .  

This still needs to have tests written.

### CODE REVIEW

**Comment by meawoppl** [comment]
Let me know if this format/implementation looks agreeable, then I can go write the tests, which should be pretty straightforward.

**Comment by pitrou** [comment]
Yes, that looks fine. I was also wondering if not passing a signature should return the whole dict of signatures to assembly.

**Comment by seibert** [comment]
+1 for a default to return the whole dict.  That is a common use case.

**Comment by meawoppl** [comment]
Can do.  I will add those this PM.  Just to clarify, you want both inspect_llvm() and inspect_asm() to return 
{
sig 0: asm 0,
...
sig N: asm N
}
When given no arguments?

**Comment by seibert** [comment]
Yes, that will be helpful for display tools (@jayvius is working on one now) that want to show all the different implementations of a function.

**Comment by seibert** [comment]
Also, now that we have merged the updated documentation, this feature should get an entry in `docs/source/reference/compilation.rst` under the `Dispatcher` documentation.

**Comment by meawoppl** [comment]
Alright, the version with the full-dict return is in the last couple commits.  It works well in my (very limited test script) which reads like the following:
```python
from numba import jit
from numba.compiler import compile_isolated
import numpy as np

@jit
def addr(a, b):
    return a + b

biggish = np.arange(1e6)

print(addr(1, 2))
print(addr(1.0, 2))

for signature in addr.signatures:
    print("LLVM:")
    print(addr.inspect_llvm(signature))
    print("\n\n\n")
    print("ASM:")
    print(addr.inspect_asm(signature))
    print("\n\n\n")

from pprint import pprint
pprint(addr.inspect_llvm())
pprint(addr.inspect_asm())
```

As suggested in the issue thread, checking for variable names in the llvm/asm is one way to check that this is working properly, others I see is checking the signature count returned after JIT'ing some known permutations of arguments, and just plain calling the function and checking that the right signatures are present.  Any other thoughts before I run off and write the testing?

**Comment by pitrou** [comment]
Hi,

> As suggested in the issue thread, checking for variable names in the llvm/asm is one way to check that this is working properly, others I see is checking the signature count returned after JIT'ing some known permutations of arguments, and just plain calling the function and checking that the right signatures are present. 

Yes, that's reasonable. Note that you shouldn't find variable names in the asm (but you'll find the function name, probably).

**Comment by meawoppl** [comment]
This is probably ready for some review me-thinks.  

This PR proved much more difficult than I appreciated as there is a number of layers involved that I wasn't even aware existed!  Is there any sort of class/architectural diagrams for this codebase?  It would seem handy for discussion if nothing else.  

As always feedback welcome.

File: docs/source/reference/compilation.rst [hunk 1]
```diff
@@ -102,6 +102,20 @@ JIT functions
 
       .. seealso:: :ref:`architecture`
 
+   .. method:: inspect_llvm(signature=None)
+
+      Return a dictionary keying compiled function signatures to the human 
+      readable llvm byte code generated for the function.  If the signature
```
Comment (pitrou) [inline docs/source/reference/compilation.rst]: Nit: I would capitalize "LLVM". Also, it's really an IR (or intermediate representation); "human readable byte code"s sounds a bit contradictory.

File: docs/source/reference/compilation.rst [hunk 2]
```diff
@@ -102,6 +102,20 @@ JIT functions
...
+
+      Return a dictionary keying compiled function signatures to the human 
+      readable llvm byte code generated for the function.  If the signature 
+      keyword is specified a string corresponding to that individual 
+      signature is returned.  
+
+   .. method:: inspect_asm(signature=None)
+
+      Return a dictionary keying compiled function signatures to the human 
+      assembler code generated for the function.  If the signature
```
Comment (pitrou) [inline docs/source/reference/compilation.rst]: I would say "human-readable native assembler code".

File: numba/dispatcher.py [hunk 3]
```diff
@@ -156,6 +156,20 @@ def _compile_for_args(self, *args, **kws):
...
+        if signature is not None:
+            lib = self._compileinfos[signature].library
+            return lib.get_llvm_str()
+
+        return dict((sig, self.inspect_llvm(sig)) for sig in self.signatures)
+
+    def inspect_asm(self, signature=None):
+        if signature is not None:
+            lib = self._compileinfos[signature].library
+            return lib.get_llvm_str()
```
Comment (pitrou) [inline numba/dispatcher.py]: This will get you the LLVM IR, not the native assembler.

File: numba/tests/test_dispatcher.py [hunk 4]
```diff
@@ -179,6 +181,55 @@ def foo(x):
...
+        foo(1.0, 1.0)
+
+        # base call to get all llvm in a dict
+        llvms = foo.inspect_llvm()
+        self.assertEqual(len(llvms), 3)
+
+        # make sure the function name shows up in the llvm
+        for llvm_bc in llvms.values():
+            # Look for the function name
+            self.assertTrue("foo" in llvm_bc)
```
Comment (pitrou) [inline numba/tests/test_dispatcher.py]: `self.assertIn(...)` would produce better error messages.

File: numba/tests/test_dispatcher.py [hunk 5]
```diff
@@ -179,6 +181,55 @@ def foo(x):
...
+        asms = foo.inspect_asm()
+        self.assertEqual(len(asms), 3)
+
+        # make sure the function name shows up in the llvm
+        for asm in asms.values():
+            # Look for the function name
+            self.assertTrue("foo" in asm)
+
+            # Look for the argument names
+            self.assertTrue("explicit_arg1" in asm)
```
Comment (pitrou) [inline numba/tests/test_dispatcher.py]: Actually, you shouldn't find the argument names in the native assembler code (try it, using NUMBA_DUMP_ASSEMBLY :-)). Althought that might depend on the platform...

**Comment by pitrou** [comment]
I left a few comments. To answer your question:

>  Is there any sort of class/architectural diagrams for this codebase? It would seem handy for discussion if nothing else. 

No, there isn't. We recently merged the new documentation so there is a slightly better developer doc than before, but it doesn't have that level of comprehensivity: http://numba.pydata.org/numba-doc/dev/developer/index.html

(note I'm very bad at drawing diagrams, myself :-))

File: numba/dispatcher.py (hunk 3 above)
Comment (meawoppl) [inline numba/dispatcher.py]: Derp.  Ty.

File: numba/tests/test_dispatcher.py (hunk 4 above)
Comment (meawoppl) [inline numba/tests/test_dispatcher.py]: noted.

File: docs/source/reference/compilation.rst (hunk 1 above)
Comment (meawoppl) [inline docs/source/reference/compilation.rst]: I didn't get to this one yet.  I said human-readable to emphasize that it was the human friendly version of the bytecode, as opposed to the usual thing you find in a .bc file.

File: docs/source/reference/compilation.rst (hunk 2 above)
Comment (meawoppl) [inline docs/source/reference/compilation.rst]: Will do tomorrow AM.

**Comment by meawoppl** [comment]
There is two doc tid bits to do, which I will hit in the AM tomorrow.

**Comment by pitrou** [comment]
Thanks for the changes! I'm merging now.

**Comment by meawoppl** [comment]
Woo woo.  Ty.

**Pull request merged**